pmpro-cancel-on-next-payment-date.php

```diff
@@ -14,94 +14,156 @@
 
-	// Are we on the cancel page and cancelling a level?
-	if ( $level == 0 && ( is_page( $pmpro_pages['cancel'] ) || ( is_admin() && ( empty( $_REQUEST['from'] ) || $_REQUEST['from'] != 'profile' ) ) ) ) {
-		// Default to false. In case we're changing membership levels multiple times during this page load.
-		$pmpro_next_payment_timestamp = false;
+	// Bypass if not level 0.
+	if ( 0 !== (int) $level ) {
+		return $level;
+	}
+	
+	// Bypass if cancelling due to an error, e.g. a payment was missed.
+	if ( 'error' === $old_level_status ) {
+		return $level;
+	}
 
-		// Get the last order.
-		$order = new MemberOrder();
-		$order->getLastMemberOrder( $user_id, 'success', $cancel_level );
+	$is_on_cancel_page = is_page( $pmpro_pages['cancel'] );
+	$is_on_profile_page = is_admin() && ( ! empty( $_REQUEST['from'] && 'profile' === $_REQUEST['from'] ) );
 
-		// Get level to check if it already has an end date.
-		if ( ! empty( $order ) && ! empty ( $order->membership_id ) ) {
-			$check_level = $wpdb->get_row( "SELECT * FROM $wpdb->pmpro_memberships_users WHERE membership_id = '" . $order->membership_id . "' AND user_id = '" . $user_id . "' ORDER BY id DESC LIMIT 1" );
-		}
+	// Bypass if not on cancellation page or a non-profile admin page.
+	// Webhook IPN calls that go through admin-ajax are non-profile admin pages.
+	if ( ! $is_on_cancel_page && ( ! is_admin() || $is_on_profile_page ) ) {
+		return $level;
+	}
 
-		// Figure out the next payment timestamp.
-		if ( empty( $check_level ) || ( ! empty( $check_level->enddate ) && $check_level->enddate != '0000-00-00 00:00:00' ) ) {
-			// Level already has an end date. Set to false so we really cancel.
-			$pmpro_next_payment_timestamp = false;
-		} elseif ( 'error' === $old_level_status ) {
-			// There was an error which led to cancellation, go ahead and cancel right away.
+	// Default to false. In case we're changing membership levels multiple times during this page load.
+	$pmpro_next_payment_timestamp = false;
+
+	// Get the last order.
+	$order = new MemberOrder();
+	$order->getLastMemberOrder( $user_id, 'success', $cancel_level );
+
+	// Get level to check if it already has an end date.
+	if ( ! empty( $order ) && ! empty( $order->membership_id ) ) {
+		$check_level = $wpdb->get_row(
+			$wpdb->prepare( "
+					SELECT *
+					FROM `{$wpdb->pmpro_memberships_users}`
+					WHERE
+						`membership_id` = %d
+						AND `user_id` = %d
+					ORDER BY `id` DESC
+					LIMIT 1
+				",
+				$order->membership_id,
+				$user_id
+			)
+		);
+	}
+
+	// Figure out the next payment timestamp.
+	if ( empty( $check_level ) || ( ! empty( $check_level->enddate ) && '0000-00-00 00:00:00' !== $check_level->enddate ) ) {
+		// Level already has an end date. Set to false so we really cancel.
+		$pmpro_next_payment_timestamp = false;
+	} elseif ( 'error' === $old_level_status ) {
+		// There was an error which led to cancellation, go ahead and cancel right away.
+		$pmpro_next_payment_timestamp = false;
```

Comment:
We can remove these lines since the bail is now separated above.

```suggestion
```

Reply:
Oh I missed that. Thanks.